### DISCUSSION TITLE
Use FileSystemProvider or Custom URI Scheme for API interactions?

### DISCUSSION THREAD

**Discussion by Jakobud** [microsoft/vscode-discussions#217] [Extension Development QnA] [answered]
I have a custom http read/write API that has some data in it. For examples sake lets just say the data is plain text data. I am able to read the data and also write data back to it.

I'm trying to write an extension that interacts with this API. What I'm interested in doing is creating a custom view in one of the sidebars that has uses a custom Tree View that would show the different endpoints.

Then you could click an endpoint to load the endpoints text data into a text editor. Make changes to it. Then when you save, it would write the data back to the API. So the text data wouldn't really exist locally in any file. Only in the text editor.

I'm trying to figure out how to do this. I'm looking through a lot of the vscode extension samples. The Tree View part isn't too difficult to figure out. But I'm not sure about how to do the rest.

I think I need to use a FileSystemProvider with custom read/write methods that would read/write the API instead of a local file. And then I guess I would need to create some sort of custom URI Scheme for it.

I've been looking at the tree-view-sample code (https://github.com/microsoft/vscode-extension-samples/) at the FileExplorer code. It has a bunch of read/write/delete/etc methods in the FileSystemProvider class but none of them seem to be used when you open or save a file via this custom File Explorer Tree.

Can anyone give me some insight here or point to examples I could check out?

**Comment by r3m0t** [comment, marked as answer]
When they click on an item in the TreeView, it should run a custom command. That command should execute the `vscode.open` command with the custom URI scheme and path, to open the editor.

After that the FS provider is used for the reading and saving back features.

https://code.visualstudio.com/api/references/commands

You could try extending this example by adding the tree view. https://github.com/microsoft/vscode-extension-samples/blob/main/fsprovider-sample/README.md

Remember you can use the `query` and `fragment` fields of the URI to encode server details, endpoint etc. You can use  `path` as well. The file extension on `path` should match the target type eg `.json`.

**Reply by Jakobud** [reply]
Okay, so the `vscode.open` command though, how do I use that to open something that isn't a local file? Really it's not "opening" anything. It needs to make an API call to a remote http server, fetch the data and then the data (which is just plain text essentially) would open in the editor.

And then when the editor is "saved" it doesn't actually save a local file but rather it sends the data back to the API.

This is one of the areas I'm not sure how to approach.

Also thanks for the link to the fsprovider. I didn't see that but it should help.

**Reply by r3m0t** [reply]
The `vscode.open` command takes a `vscode.Uri`, you just create a Uri using `vscode.Uri.parse`. The command doesn't open the file, it just tells the window to create a new tab with a new editor. You can try this by calling it for a non-existent file - the tab will still open, it will just display an error message. It's then the editor which reads the contents of the file or virtual file, and it does that using the `uri.scheme`. If the scheme is `file`, it goes to the filesystem, otherwise, it goes to an extension. When your extension activates, it calls `registerFileSystemProvider` and passes the scheme that the extension supports (knows how to read and write to). So, the editor will know to call into your extension.
